matching/README.md

```diff
@@ -0,0 +1,991 @@
+
+def compiledDefault(Matrix m, Column j): DT?
+  val m' = default(m, j)
+  if m' is None:
+    return None
+  return compile(m')
+```
+
+Note that we define a wildcard pattern as a variable pattern, or an as pattern
+whose child pattern is a wildcard pattern.
+
+`foldRanges`, `decomposeList`, `decomposeMap`, and `decomposeSet` are described
+as part of the sections on list, map, and set matching below.
+
+### Signature function
+
+`Signature` is a function that takes a column j of a matrix and a sort and
+returns the set of constructors that must be considered when pattern matching
+examines the term correpsonding to o_j.
+
+First we define `Signature` on patterns:
+
+* signature(X : Variable) = []
+* signature(c(p_1, ..., p_n)) = [c]
+* signature(p as X) = signature(p)
+* signature(List p_1, ..., p_n) = [Len(0), Len(1), ..., Len(n)]
+* signature(List p_1, ..., p_n; L; q_1, ..., q_m) = [Len(0), Len(1), ..., Len(n+m)]
+* signature(Set p_1, ..., p_n) = [Empty] if n == 0
+* signature(Set p_1, ..., p_n) = [Key(p_1), ..., Key(p_n)] if n > 0
+* signature(Set p_1, ..., p_n S) = [Key(p_1), ..., Key(p_n)]
+* signature(Map p_1 |-> q_1, ..., p_n |-> q_n) = [Empty] if n == 0
+* signature(Map p_1 |-> q_1, ..., p_n |-> q_n) = [Key(p_1), ..., Key(p_n)] if n > 0
+* signature(Map p_1 |-> q_1, ..., p_n |-> q_n M) = [Key(p_1), ..., Key(p_n)]
+
+We now define signature for columns in pseudocode:
+
+```
+def signature(Sort s, Column j): [Constructor]
+  val sigma = union from i = 1 to n of signature(P_ij)
+  if s is a map or set sort:
+    if sigma contains Empty:
+      return [Empty]
+    else:
+      return [bestKey(column j)]
+  else:
+    return sigma
+```
+
+`bestKey` will be defined later in the section for heuristics, but
+essentially, it can be known to return either `Choice` or `Key(k)` for some
+Key(k) in sigma.
+
+### Specialize function
+
+`Specialize` is a function that takes a column j of a matrix, a constructor c,
+and a matrix m and returns a decomposed matrix representing the matrix after
+having determined that the term at position o_j has constructor c.
+
+`Specialize` operates by taking each row i of the matrix and either deleting it,
+or transforming it into a new row i', depending on the structure of the pattern
+P_ij.
+
+We define specialize first on rows. We denote `a` as the arity of `c`.
+
+* specialize(c, i, X) = 
+  [P_i1, ..., P_i(j-1), Y_1, ..., Y_a, P_i(j+1), ..., P_im] -> C
+  where Y_1, ..., Y_a are fresh anonymous variables that do not appear in the
+  substitution S, and C is the clause of row i except S_i is updated with the
+  binding X |-> (o_j, s_j) if X is not an anonymous variable.
+* specialize(c, i, c(p_1, ..., p_a)) =
+  [P_i1, ..., P_i(j-1), p_1, ..., p_a, P_i(j+1), ..., P_im] -> C_i
+* specialize(c, i, c'(p_1, ..., p_a')) where c != c' = None
+* specialize(c, i, p as X) = specialize(c, i, p) except S_i is updated with the
+  binding X |-> (o_j, s_j)
```

Comment:
Does this differ from Maranget's specialisation in anything other than the tracking of variable bindings?

Reply:
No.